Make Literals foldable, ensure Parquet predicates pushdown

chris-twiner opened the pull request:
The rule actually works at this plan level, so no extension required.   At time of writing 3.x should support the getPushDowns approach, 3.5 snaps currently do at least.

I've also added a dependency on naked fs, to allow windows dev without winutils.  The StreamingFS class should go away with [#5 on naked fs](https://github.com/globalmentor/hadoop-bare-naked-local-fs/pull/5) 

Closes #343

chris-twiner commented:
nb the patch check is due to [previously untested code](https://app.codecov.io/gh/typelevel/frameless/commit/3bbdb9cac27bd71700f4c41b548719510309a02d/blob/dataset/src/main/scala/frameless/functions/package.scala#L130) being seen as new code because of the "," added :)

chris-twiner commented on dataset/src/test/scala/frameless/optimiser/LitTests.scala:
```diff
@@ -0,0 +1,44 @@
+    val dataset = TypedDataset.createUnsafe[X1[SQLTimestamp]](session.read.parquet("./target/testData"))
+
+    {
+      val pushDowns = getPushDowns(dataset.filter(dataset('a) >= ms))
```
> It's possible to use `dss.queryExecution.optimizedPlan.collect` to unroll the Expressions tree in order not to operate with strings!
> 
> `dss.queryExecution.optimizedPlan.collect { case Filter(condition, _) => condition }`
> 
> `condition` will be an Expression, in this case smth like ~ `And(not null expr, GreaterThatOrEqual)`

It's the one way to get the filters without reflection.  Happy to use that instead so we can use types.

chris-twiner replied:
per below, reflection approach used to get that actual source predicates used

chris-twiner commented:
re Databricks cluster config, it requires an uber + probably shaded jar, effectively making frameless part of databricks.  I do exactly this in [Quality](https://sparkutils.github.io/quality/0.0.2/getting_started/#configuring-on-databricks-runtimes) which isn't so straight forward.

Frameless _could_ offer a shaded/uber jar for this purpose but I think for the number of users it's better off being left in their hands and document that they have provided build dependencies in that case.  The experimental (for like 4 years now I think) optimiser route is definitely easier to integrate.

If more optimisations get added it could be worthwhile.  I did think of one whilst doing tests - struct push downs, you'd have to unpack each struct field for an entire tree to do comparisons, maps wouldn't work either.  That's [definitely fun code](https://github.com/sparkutils/quality/blob/main/src/main/scala/com/sparkutils/quality/impl/extension/AsymmetricFilterExpressions.scala) as well.

pomadchin commented on dataset/src/test/scala/frameless/optimiser/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+  }
+}
+
+class ExperimentalLitTests extends TypedDatasetSuite with TheTests {
```
I'll work a bit on the shape of a test, want to make it less self contained, more reusable.

pomadchin commented on dataset/src/test/scala/frameless/optimiser/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+package frameless.optimiser
```
`optimiser` => `optimizer`

But I want to think about the package names, will try to think about a better package structure / object names later today / this week!

chris-twiner replied:
sorry, too English, can z it up for sure whilst dropping logging....

pomadchin replied:
@chris-twiner No needs to apologize! You've done a tremendous work here 🔥

pomadchin commented on dataset/src/main/scala/frameless/optimiser/Extension.scala:
```diff
@@ -0,0 +1,10 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SparkSessionExtensions
+
+class FramelessExtension extends ((SparkSessionExtensions) => Unit) with Logging {
```
It looks like `Logging` is not used

pomadchin commented:
> So whilst most types will work with the experimental approach, structs don't, and I assume others will also not, the experimental rules don't occur early enough - a painful lesson learnt on Quality. (added verification tests for this but the reason is the Literal swap happens after the cast from X1[X4.. to struct would occur and the cast simplification rules aren't yet hit).

I think it should work, smth else is off. 

If you replace `X1[X4` with smth like:

```scala
case class Inner(a: Int, b: Int, c: Int, d: Int)
case class Test(a: Inner)
```

The output will be:

```scala
== Parsed Logical Plan ==
Filter (a#9.a > 0)
+- Relation [a#9] parquet

== Analyzed Logical Plan ==
a: struct<a:int,b:int,c:int,d:int>
Filter (a#9.a > 0)
+- Relation [a#9] parquet

== Optimized Logical Plan ==
Filter (isnotnull(a#9.a) AND (a#9.a > 0))
+- Relation [a#9] parquet

== Physical Plan ==
*(1) Filter (isnotnull(a#9.a) AND (a#9.a > 0))
+- *(1) ColumnarToRow
   +- FileScan parquet [a#9] Batched: true, DataFilters: [isnotnull(a#9.a), (a#9.a > 0)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/Users/grigory.pomadchin/subversions/git/github/pomadchin/framele..., PartitionFilters: [], PushedFilters: [IsNotNull(a.a), GreaterThan(a.a,0)], ReadSchema: struct<a:struct<a:int,b:int,c:int,d:int>>
```

chris-twiner commented:
That's not the same test though, that's a single nested field being compared.

pomadchin commented:
@chris-twiner oh, yea, I don't believe pushdown works not for primitives.

pomadchin commented:
https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L227-L235

chris-twiner commented:
perhaps not generally but they are possible to push down: [the test](https://github.com/chris-twiner/frameless/blob/temp/343_LitRule/dataset/src/test/scala/frameless/optimizer/LitTests.scala#LL107C9-L107C9)

The source.EqualTo is there when using the extension - not with experimental.

pomadchin commented:
@chris-twiner yes, [EqualTo pushdown is supported](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala#L682), but [not for all its arguments types](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala#L611-L633)

TBH (that's about naming), I think those two features are considered to be _experimental_ since Spark 2.x (don't know a specific version) :D

cchantep commented on dataset/src/test/scala/frameless/LitTests.scala:
```diff
@@ -79,7 +79,11 @@ class LitTests extends TypedDatasetSuite with Matchers {
-    ds.withColumnReplaced('alias, functions.litValue(someIpsum)).
+    val lit = functions.litValue(someIpsum)
+    val tds = ds.withColumnReplaced('alias, functions.litValue(someIpsum))
+    tds.queryExecution.toString() should include (lit.toString)
```
```suggestion

    tds.queryExecution.toString() should include (lit.toString)
```

cchantep commented on dataset/src/test/scala/frameless/GroupByTests.scala:
```diff
@@ -21,7 +21,7 @@ class GroupByTests extends TypedDatasetSuite {
       val datasetSumByA = dataset.groupByMany(A).agg(sum(B)).collect().run.toVector.sortBy(_._1)
       val sumByA = data.groupBy(_.a).map { case (k, v) => k -> v.map(_.b).map(widen).sum }.toVector.sortBy(_._1)
-
+dataset.show().run()
```
debug

chris-twiner replied:
this one was accidental, the one in lit tests was not, both removed

cchantep commented on dataset/src/main/scala/frameless/optimizer/Extension.scala:
```diff
@@ -0,0 +1,11 @@
+class FramelessExtension extends ((SparkSessionExtensions) => Unit) with Logging {
+  override def apply(extensions: SparkSessionExtensions): Unit = {
+    extensions.injectOptimizerRule( _ => LiteralRule)
+    logInfo("injected OptimizerRule - frameless.optimizer.LiteralRule")
```
```suggestion

    logInfo("Injected OptimizerRule - frameless.optimizer.LiteralRule")
```

cchantep commented on dataset/src/main/scala/frameless/functions/package.scala:
```diff
@@ -45,8 +45,8 @@ package object functions extends Udf with UnaryFunctions {
-          toCatalyst = encoder.toCatalyst(expr).genCode(_),
-          show = () => value.toString
+          show = () => value.toString,
+          encoder.toCatalyst(expr)
```
```suggestion
          catalystExpr = encoder.toCatalyst(expr)
```

cchantep commented on dataset/src/test/scala/frameless/TypedDatasetSuite.scala:
```diff
@@ -1,20 +1,33 @@
+  /**
+   * Allows bare naked to be used instead of winutils for testing / dev
+   */
+  def registerFS(sparkConf: SparkConf): SparkConf =
```
```suggestion
  def registerFS(sparkConf: SparkConf): SparkConf = {
```

cchantep commented on dataset/src/test/scala/frameless/optimizer/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+
+  import Job.framelessSparkDelayForJob
+
+  def withoutOptimisation[A]( thunk : => A): A
```
```suggestion
  def withoutOptimisation[A](thunk : => A): A
  
```

cchantep commented on dataset/src/test/scala/frameless/optimizer/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+      val dataset = TypedDataset.createUnsafe[X1[A]](session.read.parquet("./target/optimiserTestData"))
+
+      val ds = dataset.filter(op(dataset('a) ))
+      ds.explain(true)
```
```suggestion

      ds.explain(true)
      
```

cchantep commented on docs/FeatureOverview.md:
```diff
@@ -649,3 +649,66 @@ withCityInfo.select(
+  .setMaster("local[*]")
+  .setAppName("test")
+  .set("spark.ui.enabled", "false")
+sparkConfig.set("spark.sql.extensions", classOf[FramelessExtension].getName)
```
```suggestion

sparkConfig.set("spark.sql.extensions", classOf[FramelessExtension].getName)
```

cchantep commented on docs/FeatureOverview.md:
````diff
@@ -649,3 +649,66 @@ withCityInfo.select(
+  .getOrCreate()
+```
+
+FramelessExtension injects the LiteralRule optimizer rule which unpacks FramelessLit into a Spark Literal without affecting encoding.  This can have large performance benefits when enabled via extensions.
````
```suggestion
`FramelessExtension` injects the `LiteralRule` optimizer rule which unpacks `FramelessLit` into a Spark Literal without affecting encoding. This can have large performance benefits when enabled via extensions.
```

cchantep commented on docs/FeatureOverview.md:
```diff
@@ -649,3 +649,66 @@ withCityInfo.select(
+
+cp /dbfs/FileStore/uploaded_frameless.jar /databricks/jars/frameless.jar
+"""
+import java.io._
```
```suggestion

import java.io._
```

cchantep commented on docs/FeatureOverview.md:
```diff
@@ -649,3 +649,66 @@ withCityInfo.select(
+
+### experimental.extraOptimizations
+
+Using FramelessExtension will yield the best results for all Literal types as it takes place early enough to benefit from other Spark optimization rules.
```
```suggestion
Using `FramelessExtension` will yield the best results for all `Literal` types as it takes place early enough to benefit from other Spark optimization rules.
```

cchantep commented on dataset/src/test/scala/frameless/optimizer/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+
+  override implicit def session: SparkSession = s
+
+  def withoutOptimisation[A]( thunk : => A): A =
```
```suggestion
  def withoutOptimisation[A](thunk : => A): A =
```

cchantep commented on dataset/src/test/scala/frameless/optimizer/LitTests.scala:
```diff
@@ -0,0 +1,172 @@
+class ExperimentalLitTests extends TypedDatasetSuite with TheTests {
+  val isExperimental = true
+
+  def withoutOptimisation[A]( thunk : => A) = thunk
```
```suggestion
  def withoutOptimisation[A](thunk : => A) = thunk
```

chris-twiner commented on dataset/src/main/scala/frameless/functions/Lit.scala:
```diff
@@ -8,8 +8,8 @@ import org.apache.spark.sql.types.DataType
-    toCatalyst: CodegenContext => ExprCode,
-    show: () => String
+    show: () => String,
+    catalystExpr: Expression // must be the same toCatalyst as the toCatalyst function
```
updated, given it's private and only constructed in the correct way I can also just remove it

pomadchin commented:
I slightly refactored the code and tests, check [2a83510](https://github.com/typelevel/frameless/pull/721/commits/2a83510eb7972c904d01e67e46e535fa46be48f9), will work more on docs later, wondering if it makes sense to add more comprehensive DB docs 🤔

chris-twiner commented:
> I slightly refactored the code and tests, let me know what you think of [2a83510](https://github.com/typelevel/frameless/commit/2a83510eb7972c904d01e67e46e535fa46be48f9), will work more on docs later, wondering if it makes sense to add more comprehensive DB docs 🤔

looks great to me.  I added the struct tests back in.

cchantep commented on dataset/src/test/scala/frameless/sql/rules/DisabledFramelessLitRuleTests.scala:
```diff
@@ -0,0 +1,32 @@
+
+    noOptimizationRulesTest[Instant](
+      expectedStructure,
+      { case e@expressions.GreaterThanOrEqual(_, _: Lit[_]) => e },
```
```suggestion
      { case e @ expressions.GreaterThanOrEqual(_, _: Lit[_]) => e },
```

cchantep commented on dataset/src/test/scala/frameless/sql/rules/SQLRulesSuite.scala:
```diff
@@ -0,0 +1,60 @@
+    }
+  }
+
+  def noOptimizationRulesTest[A: TypedEncoder : CatalystOrdered](
```
```suggestion
  def noOptimizationRulesTest[A: TypedEncoder: CatalystOrdered](
```

pomadchin commented:
Btw, it's actually unclear why behavior in the struct tests differs, mb some optimization rule is applied after the injection. It is not just a cast though.

pomadchin commented:
1. [extra optimizations phase is at the very bottom](https://github.com/apache/spark/blob/4ffa2722d13593e46ef1c8fd59a20c439a535f61/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala#L92)
2. [injection is at the very top, before the pushdown filter prunning](https://github.com/apache/spark/blob/4ffa2722d13593e46ef1c8fd59a20c439a535f61/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala#L50)
  * It is in the [extendedOperatorOptimizationRules](https://github.com/apache/spark/blob/4ffa2722d13593e46ef1c8fd59a20c439a535f61/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L128)
  
I know for sure that it's not `SimplifyCasts`, because types don't match. We could add our own, a more relaxed SimplifyCasts for the `Struct(not nullable fields)` on the left, and `Struct(nullable fields)` on the right, they are definitely coercible. All not nullables can be converted to nullables, and that's what `cast` wants, because of the not simplified cast this literal is not pushed down.

chris-twiner commented:
> Btw, it's actually unclear why behavior in the struct tests differs, mb some optimization rule is applied after the injection. It is not just a cast though.

Challenge accepted!!!

Unlike the Quality case this can be fixed :)

The rules are added to [defaultBatches](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala#L92) and added to [Optimizer.defaultBatches which adds ConstantFolding](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L103). 

As was:
[ConstantFolding](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala#L63) will enter but because in (a#9 = cast(FramelessLit(X4(1,2,3,4)) as struct<a:int,b:int,c:int,d:int>)) neither the attributeref nor the cast+Lit are foldable line 80 won't trigger.

So, adding foldable = true to Lit gets the tests passing.

But wait!! There's more!!  Given folding requires a usable eval, compilation is too early and would be triggered at least twice anyway.  I fixed this for UDF's before but missed the pattern on Lit completely it seems.  For UDF it makes sense to pre-compile. I don't think it does for Lit's, particularly when we are attempting to optimise them out.

Now to the really fun stuff, the rule had:

```scala
case Lit(dataType, _, _, convertedValue) => Literal(convertedValue.eval(), dataType)
```

but the ConstantFolding had:

```scala
Literal.create(e.eval(EmptyRow), e.dataType)
```

which kinda looks the same.  If you've made it this far, dear reader, you'll no doubt realise we don't need the rule and foldable = true was the real hero all along!

I've pushed the working tests with the rule thoroughly disabled.  If you are ok with me removing the rule and docs I'll happily do it :)  Should the config prep stuff also be removed as the two tests are effectively identical now?

chris-twiner commented:
nb - we could have foldable ask catalystExpr if it's foldable, but given Literal doesn't and these are meant to be values it should be fine.

pomadchin commented on dataset/src/main/scala/frameless/functions/Lit.scala:
```diff
@@ -8,52 +8,18 @@ import org.apache.spark.sql.types.DataType
 ) extends Expression with NonSQLExpression {
   override def toString: String = s"FramelessLit(${show()})"
 
-  def eval(input: InternalRow): Any = {
-    val ctx = new CodegenContext()
-    val eval = genCode(ctx)
-
-    val codeBody = s"""
-      public scala.Function1<InternalRow, Object> generate(Object[] references) {
-        return new LiteralEvalImpl(references);
-      }
-
-      class LiteralEvalImpl extends scala.runtime.AbstractFunction1<InternalRow, Object> {
-        private final Object[] references;
-        ${ctx.declareMutableStates()}
-        ${ctx.declareAddedFunctions()}
-
-        public LiteralEvalImpl(Object[] references) {
-          this.references = references;
-          ${ctx.initMutableStates()}
-        }
-
-        public java.lang.Object apply(java.lang.Object z) {
-          InternalRow ${ctx.INPUT_ROW} = (InternalRow) z;
-          ${eval.code}
-          return ${eval.isNull} ? ((Object)null) : ((Object)${eval.value});
-        }
-      }
-    """
-
-    val code = CodeFormatter.stripOverlappingComments(
-      new CodeAndComment(codeBody, ctx.getPlaceHolderToComments())
-    )
-
-    val (clazz, _) = CodeGenerator.compile(code)
-    val codegen =
-      clazz.generate(ctx.references.toArray).asInstanceOf[InternalRow => AnyRef]
-
-    codegen(input)
-  }
```
I think this code was left to deal with some serialization issues long time ago. 

I was able to find the old commit that introduced it: https://github.com/typelevel/frameless/blame/59caada7955c90fce266577109122265b7840418/dataset/src/main/scala/frameless/functions/Lit.scala 

Let's not remove this code, for now?

pomadchin replied:
Serialization test failures could be tricky to find, i'd vote for a separate PR with investigations ~ mb indeed we don't need this portion of code anymore.

pomadchin commented:
@chris-twiner it looks like spark-32 dataset tests are failing, I think it is fine until it works in Spark 3.3+, but the CI is unhappy.

pomadchin commented:
We could also keep the rule with the extensions injector, just in case, that is required for the Spark 3.2; It is not hard to maintain and a nice demo of the optimization rules implementation. If it makes tests easier to maintain its worth it.

chris-twiner commented:
it's because the values aren't foldable:
```scala
override val foldable: Boolean = true//catalystExpr.foldable
```

In 3.3 InvokeLike introduced:
```scala
override def foldable: Boolean =
    children.forall(_.foldable) && deterministic && trustedSerializable(dataType)
```

so it's missing from 3.2 .  That stops timestamp and instant.  

```
if (isnull(X4(1,2,3,4))) null else named_struct(a, X4(1,2,3,4).a, b, X4(1,2,3,4).b, c, X4(1,2,3,4).c, d, X4(1,2,3,4).d)
````

isnull is foldable, null is foldable but named_struct is not because of the invokes.

So the rule itself isn't needed, it's equivalent to foldable = true

pomadchin commented:
@chris-twiner yea, let's make it just true in this case :+1: good to know. Lets add a link to your comments into the code comments so we don't forget the decision made!

chris-twiner commented:
> @chris-twiner yea, let's make it just true in this case 👍 good to know. Lets add a link to your comments into the code comments so we don't forget the decision made!

done, I guess github has had enough of this for a bit though, the CI isn't running.

pomadchin commented on dataset/src/main/scala/frameless/functions/Lit.scala:
```diff
@@ -8,16 +8,17 @@ import org.apache.spark.sql.types.DataType
   override def toString: String = s"FramelessLit(${show()})"
 
-  def eval(input: InternalRow): Any = {
+  lazy val codegen = {
```
Any reason for making it a lazy val and not a regular function?

chris-twiner replied:
it could get evaluated several times over potentially before a wholestage.  Well perhaps it doesn't after the foldable change.  Either way very little overhead from the lazy and much quicker than calling a second time.

pomadchin reviewed (approved):
🔥

chris-twiner commented:
Per the text in [SPARK-40380](https://issues.apache.org/jira/browse/SPARK-40380) this is probably not correct as it'd affect sparksql-scalapb a user of frameless.

Rather than break out another source compatibility layer I'll look at a compile time elide or similar and sub in a different solution for 3.2.

Question is should the solution just be accept failure on 3.2 (and document predicate pushdown won't work on 3.2) or actually provide a backport of 40380?

pomadchin commented:
@chris-twiner could you tell more, how that issue is related to this PR, and to sparksql-scalapb?

chris-twiner commented:
> @chris-twiner could you tell more, how that issue is related to this PR, and to sparksql-scalapb?

sorry, was trying to create an example.  With foldable always true Invokes could create unserializable expressions (that are ObjectType results).  These would be fine when not folded, but folding will cause the query to fail as the expression in Literal would be non-serializable.

3.3.1 and above "fixes" that by this code in InvokeLike:

```scala
override def foldable: Boolean =
    children.forall(_.foldable) && deterministic && trustedSerializable(dataType)
// Returns true if we can trust all values of the given DataType can be serialized.
  private def trustedSerializable(dt: DataType): Boolean = {
    // Right now we conservatively block all ObjectType (Java objects) regardless of
    // serializability, because the type-level info with java.io.Serializable and
    // java.io.Externalizable marker interfaces are not strong guarantees.
    // This restriction can be relaxed in the future to expose more optimizations.
    !dt.existsRecursively(_.isInstanceOf[ObjectType])
  }
```

so by not calling catalystExpr.foldable we risk introducing an unserializable ObjectType.

pomadchin commented:
@chris-twiner we could add this function into the `FramelessInternals` package:

```scala
  // https://github.com/typelevel/frameless/pull/721
  // TODO: remove with the Spark 3.2 support drop
  def isFoldableExpressionCompat(expr: Expression): Boolean = {
    // Returns true if we can trust all values of the given DataType can be serialized.
    def trustedSerializable(dt: DataType): Boolean = {
      // Right now we conservatively block all ObjectType (Java objects) regardless of
      // serializability, because the type-level info with java.io.Serializable and
      // java.io.Externalizable marker interfaces are not strong guarantees.
      // This restriction can be relaxed in the future to expose more optimizations.
      !dt.existsRecursively(_.isInstanceOf[ObjectType])
    }

    expr.children.forall(_.foldable) && expr.deterministic && trustedSerializable(expr.dataType)
  }
```

and then on the user side it is smth like 

```scala
// in Lit.scala
// https://github.com/typelevel/frameless/pull/721
// TODO: replace with catalystExpr.foldable with the Spark 3.2 drop
override val foldable: Boolean = FramelessInternals.isFoldableExpressionCompat(catalystExpr)
```

chris-twiner commented:
That will fail on 3.2 though.  It would need to be a bottom up check with that logic called for InvokeLike.

pomadchin commented:
Let's still add this backport for the 3.2/3.3+ compat, but we'll leave a note about this issue, imo that's the Spark problem/

chris-twiner commented:
> Let's still add this backport for the 3.2/3.3+ compat (we need that foldable check), but we'll leave a note about this issue, imo that's the Spark problem.

agreed, I've updated the comment on that and disabled the tests for 3.2.

pomadchin commented on dataset/src/test/scala/frameless/sql/rules/SQLRulesSuite.scala:
```diff
@@ -17,6 +17,27 @@ trait SQLRulesSuite extends TypedDatasetSuite with Matchers { self =>
     s"$tmpDir/${self.getClass.getName}"
   }
 
+  lazy val sparkFullVersion = {
+    val pos = classOf[Expression].getPackage.getSpecificationVersion
+    if (pos eq null) // DBR is always null
+      SparkSession.active.version // taking a running spark version string, hence lazy
+    else
+      pos
+  }
+
+  lazy val sparkVersion = {
+    sparkFullVersion.split('.').take(2).mkString(".")
+  }
+
+  /**
+   * Don't run this test on 3.2
+   */
+  def not3_2[T](thunk: => T): Any =
+    if (sparkVersion != "3.2")
+      thunk
+    else
+      ()
```
I'm more in favor of back-porting the function into the FramelessInternals, these hacks are harder to maintain.

chris-twiner replied:
no, I agree, just wanted a "correct" impl committing first.

pomadchin replied:
I'll let you finish it then first :D

chris-twiner replied:
moreover having a divergent impl isn't great so it'll be another compat layer....

pomadchin replied:
Yea, so just a spark cross compilable function is good enough :+1: We just need to remove it once we drop Spark 3.2

chris-twiner replied:
Ok, that last push has the goods.  There is a gotcha, in that there are probably other Expressions that don't implement foldable in 3.2 but do in later versions [searching for the obvious](https://issues.apache.org/jira/browse/SPARK-40380?jql=project%20%3D%20SPARK%20AND%20text%20~%20%22constant%20folding%22%20order%20by%20key%20desc) only shows [SPARK-40380](https://issues.apache.org/jira/browse/SPARK-40380) and [SPARK-39106](https://issues.apache.org/jira/browse/SPARK-39106) so we may be ok as I cover both.

In the case where there is another impl difference the worst that should happen now is it doesn't get pushed down.

pomadchin replied:
idk; we should not patch Spark bugs imo, I was only suggesting adding a cross compiling function that will properly work for Spark 3.3+ and will still be incorrect for the Spark 3.2, like here: https://github.com/typelevel/frameless/pull/721#issuecomment-1582729739

pomadchin commented on dataset/src/main/spark-32/frameless/FoldableImpl.scala:
```diff
@@ -0,0 +1,54 @@
+import org.apache.spark.sql.catalyst.expressions.objects.InvokeLike
+import org.apache.spark.sql.types.{DataType, ObjectType}
+
+object FoldableImpl {
```
Are you sure we need it? Why not to add an alias into `FramelessInternals`? It won't fix the Spark 3.2 issue, but will make it compile.

chris-twiner replied:
Probably not.  I'd guess it's not likely there are huge Lit expression trees that need transformUp'ing that it's worth saving the computational time.

The only risk of doing that is they change the impl in 3.3/3.4 so it's no longer correct for them, and that's probably only for InovkeLike, I can't imagine an need to change the ConditionalExpression logic.

pomadchin replied:
I'd say lets add the less invasive change, don't address possible Spark 3.2 problems, and we'll get rid of 3.2 with the 3.5 release.

chris-twiner replied:
sorry, it's been a day without much coffee, I don't understand that last bit, it seems contradictory. I seriously need more coffee but I'd like to sleep - ah the coders dilemma.

do you want either:

a) 3.2 issues are addressed by moving spark-32/frameless/FoldableImpl into FramelessInternals fully and accepting that this will be run on 3.3 and 3.4 with a possible risk of divergent implementation. < the pushdown tests run all versions and the project structure stays as per the previous release / #699 
b) we don't address 3.2 issues and so we don't need FramelessInternals as we can just call catalystExpr.foldable  < the tests should be disabled on 3.2 (but with project structure as per #699), or
c) yet something else?

pomadchin replied:
1. setting `foldable = true for Spark 3.2`, invoke `catalystExpr.foldable` for 3.3+
2. setting `foldable = false for Spark 3.2`, invoke `catalystExpr.foldable` for 3.3+, but put the test into the Spark 3.3+ test folder, similar to what we've done with `MapGroups`.

chris-twiner replied:
ok, 1 has the risk of failure per [SPARK-40380](https://issues.apache.org/jira/browse/SPARK-40380) so I can't recommend that.

I'd offer a 3 in that case.:

use catalystExpr.foldable (equates to false on 3.2 but is fully compatible behaviour wise) and split the tests.

pomadchin replied:
Yea, makes sense :+1: So later on we just remove the tests, awesome.

pomadchin commented on build.sbt:
```diff
@@ -83,6 +86,7 @@ lazy val `dataset-spark32` = project
   .settings(name := "frameless-dataset-spark32")
   .settings(sourceDirectory := (dataset / sourceDirectory).value)
   .settings(Compile / unmanagedSourceDirectories += (dataset / baseDirectory).value / "src" / "main" / "spark-3")
+  .settings(Test / unmanagedSourceDirectories += (dataset / baseDirectory).value / "src" / "test" / "spark-32")
```
`spark-32` => `spark-3.2`

pomadchin commented on build.sbt:
```diff
@@ -74,6 +76,7 @@ lazy val `dataset-spark33` = project
   .settings(name := "frameless-dataset-spark33")
   .settings(sourceDirectory := (dataset / sourceDirectory).value)
   .settings(Compile / unmanagedSourceDirectories += (dataset / baseDirectory).value / "src" / "main" / "spark-3")
+  .settings(Test / unmanagedSourceDirectories += (dataset / baseDirectory).value / "src" / "test" / "spark-3+")
```
`spark-3+` => `spark-3.3+`

pomadchin reviewed (commented):
LGTM! Left a couple of suggestions around the dirs structure!

pomadchin reviewed (approved):
🔥 Quite a PR

chris-twiner commented:
> 🔥 Quite a PR

indeed, but a fun journey
